sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala

```diff
@@ -223,19 +223,19 @@ case class BroadcastHashJoinExec(
    * Handles NULL-aware anti join (NAAJ) separately here.
    */
   protected override def codegenAnti(ctx: CodegenContext, input: Seq[ExprCode]): String = {
-    if (isNullAwareAntiJoin) {
-      val (broadcastRelation, relationTerm) = prepareBroadcast(ctx)
+    val (broadcastRelation, relationTerm) = prepareBroadcast(ctx)
```

Comment:
After a second look, I think this is incorrect. We may go to the `else` branch, which calls `super.codegenAnti` and call `prepareBroadcast` again.

I think we should keep this method untouched, and update `HashJoin.codegenAnti` to fast stop with `EmptyHashedRelation`.

Reply:
I thought about this problem too.
HashJoin.scala does not have the API prepareBroadcast, only the following prepareRelation available, which will not return HashedRelation.

```
protected def prepareRelation(ctx: CodegenContext): (String, Boolean)
```

I was thinking that call prepareBroadcast twice is OK since broadcast.Broadcast[HashedRelation] is a future call, and the collect and build action will only happen once, but the relationTerm will be generated twice

is it ok to just change prepareRelation to return a (String, Boolean, HashedRelation) ??
tell me when you decided which way to go. @cloud-fan

Reply:
I was also consider doing it inside BHJ because in ShuffledHashJoinExec
the hashedRelation is built among generated-code, it might be hard to just return hashedRelation in prepareRelation call.

```
protected override def prepareRelation(ctx: CodegenContext): (String, Boolean) = {
    val thisPlan = ctx.addReferenceObj("plan", this)
    val clsName = classOf[HashedRelation].getName

    // Inline mutable state since not many join operations in a task
    val relationTerm = ctx.addMutableState(clsName, "relation",
      v => s"$v = $thisPlan.buildHashedRelation(inputs[1]);", forceInline = true)
    (relationTerm, false)
  }
```